R/expr-string.R

```diff
@@ -0,0 +1,1158 @@
+#' @param dtype The data type to convert into. Can be either `pl$Date`,
+#' [`pl$Datetime()`][DataType_Datetime], or `pl$Time`.
+#' @param format Format to use for conversion. Refer to
+#' [the chrono crate documentation](https://docs.rs/chrono/latest/chrono/format/strftime/index.html)
+#' for the full specification. Example: `"%Y-%m-%d %H:%M:%S"`.
+#' If `NULL` (default), the format is inferred from the data.
+#' Notice that time zone `%Z` is not supported and will just ignore timezones.
+#' Numeric time zones like `%z` or `%:z` are supported.
+#' @param ... Not used.
+#' @param strict If `TRUE` (default), raise an error if a single string cannot
+#' be parsed. If `FALSE`, produce a polars `null`.
+#' @param exact If `TRUE` (default), require an exact format match. If `FALSE`,
+#' allow the format to match anywhere in the target string.
+#' Conversion to the Time type is always exact.
+#' Note that using `exact = FALSE` introduces a performance penalty -
+#' cleaning your data beforehand will almost certainly be more performant.
+#' @param cache Use a cache of unique, converted dates to apply the datetime
+#' conversion.
+#' @inheritParams ExprDT_replace_time_zone
+#' @return [Expr][Expr_class] of Date, Datetime or Time type
+#' @seealso
+#' - [`<Expr>$str$to_date()`][expr_str_to_date]
+#' - [`<Expr>$str$to_datetime()`][expr_str_to_datetime]
+#' - [`<Expr>$str$to_time()`][expr_str_to_time]
+#' @examples
+#' # Dealing with a consistent format
+#' s <- as_polars_series(c("2020-01-01 01:00Z", "2020-01-01 02:00Z"))
+#'
+#' s$str$strptime(pl$Datetime(), "%Y-%m-%d %H:%M%#z")
+#'
+#' # Auto infer format
+#' s$str$strptime(pl$Datetime())
+#'
+#' # Datetime with timezone is interpreted as UTC timezone
+#' as_polars_series("2020-01-01T01:00:00+09:00")$str$strptime(pl$Datetime())
+#'
+#' # Dealing with different formats.
+#' s <- as_polars_series(
+#'   c(
+#'     "2021-04-22",
+#'     "2022-01-04 00:00:00",
+#'     "01/31/22",
+#'     "Sun Jul  8 00:34:60 2001"
+#'   ),
+#'   "date"
+#' )
+#'
+#' s$to_frame()$select(
+#'   pl$coalesce(
+#'     pl$col("date")$str$strptime(pl$Date, "%F", strict = FALSE),
+#'     pl$col("date")$str$strptime(pl$Date, "%F %T", strict = FALSE),
+#'     pl$col("date")$str$strptime(pl$Date, "%D", strict = FALSE),
+#'     pl$col("date")$str$strptime(pl$Date, "%c", strict = FALSE)
+#'   )
+#' )
+#'
+#' # Ignore invalid time
+#' s <- as_polars_series(
+#'   c(
+#'     "2023-01-01 11:22:33 -0100",
+#'     "2023-01-01 11:22:33 +0300",
+#'     "invalid time"
+#'   )
+#' )
+#'
+#' s$str$strptime(
+#'   pl$Datetime("ns"),
+#'   format = "%Y-%m-%d %H:%M:%S %z",
+#'   strict = FALSE
+#' )
+expr_str_strptime <- function(
+    dtype,
+    format = NULL,
+    ...,
+    strict = TRUE,
+    exact = TRUE,
+    cache = TRUE,
+    ambiguous = "raise") {
+  wrap({
+    check_dots_empty0(...)
+    check_polars_dtype(dtype)
+
+    if (pl$same_outer_dt(dtype, pl$Datetime())) {
```

Comment:
Does the `pl$same_outer_dt` function exist?
Doesn't you want to use `$is_*` methods of dtypes?

Reply:
No it doesn't exist, I didn't modify those string-datetime functions too much since I'd like to address them when focusing on the date(time) expressions. However, there is no `is_datetime()` for datatype, even upstream. It shouldn't be hard to add however:
```rs
matches!(self, Datetime(_, _))
```

and actually I just realized that `is_temporal()` isn't only `Time`.

Reply:
Good point.
How about creat a PR to upstream polars?
(If you don't want to do, I will)

Reply:
I'll do it

Reply:
There was already some discussion in https://github.com/pola-rs/polars/issues/12196, I opened an issue here first: https://github.com/pola-rs/polars/issues/19235

Reply:
Seems like they have no interest for this, issue was closed. Using `$eq(pl$Datetime())` should do the trick I think